src/pages/VenuePage/VenuePage.tsx

```diff
@@ -115,92 +87,63 @@ const VenuePage: React.FC = () => {
-    ? !profile?.lastSeenIn[location]
-    : false;
+    trackLocationEntered({ userId, locationName: venueName });
+  }, [userId, venueName]);
 
-  const newLocations = {
-    ...prevLocations,
-    ...newLocation,
-  };
+  useInterval(() => {
+    updateUserLocationToCurrentVenue();
+  }, LOC_UPDATE_FREQ_MS);
 
   useEffect(() => {
-    if (
-      user &&
-      location &&
-      isNewLocation &&
-      ((!unmounted && !retainAttendance) || retainAttendance) &&
-      (!profile?.lastSeenIn || !profile?.lastSeenIn[location])
-    ) {
-      updateLocationData(
-        user,
-        location ? newLocations : prevLocations,
-        profile?.lastSeenIn
-      );
-      setUnmounted(false);
-    }
-    if (
-      user &&
-      profile &&
-      (profile.lastSeenIn === null || profile?.lastSeenIn === undefined)
-    ) {
-      updateLocationData(
-        user,
-        location ? newLocations : prevLocations,
-        profile?.lastSeenIn
-      );
-    }
-  }, [
-    isNewLocation,
-    location,
-    newLocation,
-    newLocations,
-    prevLocations,
-    profile,
-    retainAttendance,
-    unmounted,
-    user,
-  ]);
+    updateUserLocationToCurrentVenue();
+  }, [updateUserLocationToCurrentVenue]);
+
+  // useEffect(() => {
+  //   if (!userId || !venueName) return;
+  //   updateUserLocationToCurrentVenue();
+
+  // NOTE: A suggestion on how to avoid location cleaning, when two tabs were opened and one of them was closed
+
+  // document.addEventListener("visibilitychange", () => {
+  //   if (document.visibilityState === "visible") {
+  //     updateUserLocationToCurrentVenue();
+  //   }
+  // });
+  // }, [userId, venueName, updateUserLocationToCurrentVenue]);
 
   useEffect(() => {
-    const leaveRoomBeforeUnload = () => {
-      if (user && !retainAttendance) {
-        const locations = { ...prevLocations };
-        delete locations[venueName];
-        setUnmounted(true);
-        updateLocationData(user, locations, undefined);
-      }
-    };
-    window.addEventListener("beforeunload", leaveRoomBeforeUnload, false);
-    return () => {
-      window.removeEventListener("beforeunload", leaveRoomBeforeUnload, false);
-    };
-  }, [prevLocations, retainAttendance, user, venueName]);
+    if (!userId) return;
+
+    // NOTE: Clear user location on page close
+    window.addEventListener("beforeunload", () =>
+      updateLocationData(userId, {})
+    );
+  }, [userId]);
 
   useEffect(() => {
     if (
       profile?.enteredVenueIds &&
-      venue?.id &&
-      profile?.enteredVenueIds.includes(venue?.id)
+      venueId &&
+      profile?.enteredVenueIds.includes(venueId)
     )
       return;
-    if (!venue || !user) return;
-    updateProfileEnteredVenueIds(
-      profile?.enteredVenueIds,
-      user?.uid,
-      venue?.id
-    );
-  }, [profile, user, venue]);
+    if (!venueId || !user || !profile) return;
+
+    updateProfileEnteredVenueIds(profile?.enteredVenueIds, user?.uid, venueId);
```

Comment:
This should be refactored to use proper firestore update/arrayUnion/etc logic, instead of needing to pass through the old 'enteredVenueIds' and overwriting the entire field in the database.

At that point, depending on whether we already have the `enteredVenueIds` in the frontend, we wouldn't even need to check if the current `venueId` is already there, as we could just always 'union' it, and then ensure that this function is only called when the `userId`/`venueId` changes.